launchable/test_runners/cts.py

```diff
@@ -0,0 +1,178 @@
+                    stack_trace_element = failure.find("StackTrace")
+                    if stack_trace_element is not None:
+                        stack_trace = str(stack_trace_element.text)
+
+                    stdout = failure.get("message", "")
+                    stderr = stack_trace
+
+                test_results.append(TestResult(
+                    test_case_name=test_case_name,
+                    test_name=test_name,
+                    result=result,
+                    stdout=stdout,
+                    stderr=stderr))
+
+        if len(test_results) == 0:
+            continue
+
+        test_duration_msec_per_test = int(total_duration) / len(test_results)
+
+        for test_result in test_results:  # type: TestResult
+            if module_name == "" or test_result.test_case_name == "" or test_result.test_name == "":
+                continue
+
+            events.append(CaseEvent.create(
+                test_path=build_record_test_path(module_name, test_result.test_case_name, test_result.test_name),
+                duration_secs=float(test_duration_msec_per_test / 1000),
+                status=test_result.case_event_status(),
+                stdout=test_result.stdout,
+                stderr=test_result.stderr,
+            ))
+
+    return (x for x in events)
+
+
+@click.argument('reports', required=True, nargs=-1)
+@launchable.record.tests
+def record_tests(client, reports):
+    """
+    Beta: Report test result that Compatibility Test Suite (CTS) produced
+    """
+    for r in reports:
+        client.report(r)
+
+    client.parse_func = parse_func
+    client.run()
+
+
+@launchable.subset
+def subset(client):
+    """
+    Beta: Produces test list from previous test sessions for Compatibility Test Suite (CTS)
+    """
+    start_module = False
+
+    """ # noqa: E501
+    # This is sample output of `cts-tradefed list modules`
+    ==================
+    Notice:
+    We collect anonymous usage statistics in accordance with our Content Licenses (https://source.android.com/setup/start/licenses), Contributor License Agreement (https://opensource.google.com/docs/cla/), Privacy Policy (https://policies.google.com/privacy) and Terms of Service (https://policies.google.com/terms).
+    ==================
+    Android Compatibility Test Suite 12.1_r5 (9566553)
+    Use "help" or "help all" to get more information on running commands.
+    Non-interactive mode: Running initial command then exiting.
+    Using commandline arguments as starting command: [list, modules]
+    arm64-v8a CtsAbiOverrideHostTestCases[instant]
+    arm64-v8a CtsAbiOverrideHostTestCases[secondary_user]
+    arm64-v8a CtsAbiOverrideHostTestCases
+    armeabi-v7a CtsAbiOverrideHostTestCases
+    """
+
+    for t in client.stdin():
+        if "starting command" in t:
+            start_module = True
+            continue
+
+        if not start_module:
+            continue
+
+        lines = t.rstrip("\n").split()
+        if len(lines) != 2:
```

Comment:
[question]
Is this always 2 lines?
What consists of these 2 lines?

Reply:
I want to parse these lines and get only CtsAbiOverrideHostTestCases[instant], CtsAbiOverrideHostTestCases[secondary_user]...

```
    arm64-v8a CtsAbiOverrideHostTestCases[instant]
    arm64-v8a CtsAbiOverrideHostTestCases[secondary_user]
    arm64-v8a CtsAbiOverrideHostTestCases
    armeabi-v7a CtsAbiOverrideHostTestCases
```

However, it's hard to understand..., so I'll rename the valuable name and add comments.

Reply:
[question]
So `lines = t.rstrip("\n").split()` is intended to do:
E.g. " arm64-v8a CtsAbiOverrideHostTestCases[secondary_user]\n"
- Remove "\n"
- Split by space and make a list: ['arm64-v8a', 'CtsAbiOverrideHostTestCases[secondary_user]']

Reply:
yes

Reply:
https://github.com/launchableinc/cli/pull/555/commits/6ed582deed125df52edf3850dba91fbefe9089e4

Reply:
Just to be sure, I see there are multiple test summary format depending on Android version and CTS version.
https://source.android.com/docs/compatibility/cts/interpret?hl=ja
It is intended to take Android 10 format here, correct?

Reply:
we'll support only CTS v2, for now